compiler/src/dotty/tools/backend/jvm/GenBCode.scala

```diff
@@ -68,10 +68,47 @@ class GenBCode extends Phase {
           // If we close the jar the next run will not be able to write on the jar.
           // But if we do not close it we cannot use it as part of the macro classpath of the suspended files.
           report.error("Can not suspend and output to a jar at the same time. See suspension with -Xprint-suspension.")
+
+        updateJarManifestWithMainClass(units, jar)
         jar.close()
       case _ =>
     }
   }
+
+  private def updateJarManifestWithMainClass(units: List[CompilationUnit], jarArchive: JarArchive)(using Context): Unit =
+    val mainClass = Option.when(!ctx.settings.XmainClass.isDefault)(ctx.settings.XmainClass.value).orElse {
+      val _mainClassesBuffer = new mutable.HashSet[String]
+      units.map { unit =>
+        unit.tpdTree.foreachSubTree { tree =>
+          val sym = tree.symbol
+          import dotty.tools.dotc.core.NameOps.stripModuleClassSuffix
+          val name = sym.fullName.stripModuleClassSuffix.toString
+          if (sym.isStatic && !sym.is(Flags.Trait) && ctx.platform.hasMainMethod(sym)) {
+            // If sym is an object, all main methods count, otherwise only @static ones count.
+            _mainClassesBuffer += name
+          }
+        }
+      }
+      _mainClassesBuffer.toList.match
+        case List(mainClass) =>
+          Some(mainClass)
+        case Nil =>
+          report.warning("No Main-Class designated or discovered.")
+          None
+        case mcs =>
+          report.warning(s"No Main-Class due to multiple entry points:\n  ${mcs.mkString("\n  ")}")
+          None
+    }
```

Comment:
This is copied from `compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala` I was thinking whether we could extract it for reuse rather than copy-paste, yet I wasn't sure where should I place it, maybe somewhere under `dotty.tools.dotc.util`

Reply:
Please feel free to extract any code for reuse yes, doesn't matter too much where it ends up :). However, having to fully traverse every single compilation unit just to get main method names out is going to be fairly expensive and should be avoided, especially for information that we should already know. Maybe ExtractAPI (or some other phase which is always run if we want this to work even outside of sbt) should collect this information and store it in the Context. In fact, we used to have a phase to do that which was supposed to do that but I deleted it because it was dead code: https://github.com/lampepfl/dotty/commit/3ebf91983e3a7427a3c03e9cd56ec548f51f2295

Reply:
Thanks for the advice, I will look into that :D